language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs

```diff
@@ -58,110 +56,203 @@ integerToInt64 = fromIntegral
 -- | Apply a function with a known 'TypeScheme' to a list of 'Constant's (unwrapped from 'Value's).
 -- Checks that the constants are of expected types.
 applyTypeSchemed
-    :: forall e m args uni res.
-       ( MonadError (ErrorWithCause uni e) m, AsUnliftingError e, AsConstAppError e uni
-       , SpendBudget m uni, Closed uni, uni `Everywhere` ExMemoryUsage
+    :: forall err m args term res.
+       ( MonadError (ErrorWithCause term err) m, AsUnliftingError err, AsConstAppError err term
+       , SpendBudget m term
        )
     => StagedBuiltinName
-    -> TypeScheme uni args res
+    -> TypeScheme term args res
     -> FoldArgs args res
     -> FoldArgsEx args
-    -> [Value TyName Name uni ExMemory]
-    -> m (ConstAppResult uni ExMemory)
+    -> [term]
+    -> m (ConstAppResult term)
 applyTypeSchemed name = go where
     go
         :: forall args'.
-           TypeScheme uni args' res
+           TypeScheme term args' res
         -> FoldArgs args' res
         -> FoldArgsEx args'
-        -> [Value TyName Name uni ExMemory]
-        -> m (ConstAppResult uni ExMemory)
+        -> [term]
+        -> m (ConstAppResult term)
     go (TypeSchemeResult _)        y _ args =
         -- TODO: The costing function is NOT run here. Might cause problems if there's never a TypeSchemeArrow.
         case args of
-            -- TODO is `withMemory` a good idea here?
-            [] -> pure . ConstAppSuccess $ withMemory $ makeKnown y    -- Computed the result.
+            [] -> pure . ConstAppSuccess $ makeKnown y    -- Computed the result.
             _  -> throwingWithCause _ConstAppError        -- Too many arguments.
-                    (ExcessArgumentsConstAppError $ void <$> args)
+                    (ExcessArgumentsConstAppError args)
                     Nothing
     go (TypeSchemeAllType _ schK)  f exF args =
         go (schK Proxy) f exF args
     go (TypeSchemeArrow _ schB)    f exF args = case args of
         []          -> pure ConstAppStuck                 -- Not enough arguments to compute.
         arg : args' -> do                                 -- Peel off one argument.
             -- Coerce the argument to a Haskell value.
-            x <- readKnown $ void arg
-            let budget = exF $ termAnn arg
+            x <- readKnown arg
+            exF' <- feedBudgeter exF arg
```

Comment:
Previously we needed to know that `arg` is annotated with `ExMemory`, but that doesn't fit well into the generalized version of the builtin application machinery, so I moved that logic into the `SpendBudget` class.